File: Prac/tree.cpp
#include<iostream>
#include<queue>
#include<algorithm>
using namespace std;


struct node{

    int data;
    struct node* left;
    struct node* right;

    node(int val){
        data=val;
        left = NULL;
        right = NULL;
    }

};


void preOrder(node * root){
    if(root==NULL){
        return;
    }
    cout<<root->data<<" ";
    preOrder(root->left);
    preOrder(root->right);
}

void inOrder(node * root){
    if(root==NULL){
        return;
    }
    preOrder(root->left);
    cout<<root->data<<" ";
    preOrder(root->right);
}

void postOrder(node * root){
    if(root==NULL){
        return;
    }
    preOrder(root->left);
    preOrder(root->right);
    cout<<root->data<<" ";
}



void levelOrder(node * root){

    if(root==NULL){
        return;
    }

    queue<node*> q;
    q.push(root);
    q.push(NULL);

    while(!q.empty()){
        node* n = q.front();
        q.pop();
        if(n!=NULL){
            cout<<n->data<<" ";
            if(n->left){
                q.push(n->left);
            }
            if(n->right){
                q.push(n->right);
            }
        }
        else if(!q.empty()){
                q.push(NULL);
        }
    }

}

int sumAtK(node * root,int k){
    if(root == NULL){
        return -1;
    }

    queue<node*>q;
    q.push(root);
    q.push(NULL);
    int level =0;
    int sum =0;

    while (!q.empty())
    {
        node * n = q.front();
        q.pop();
        
        if(n!=NULL){
            if(level ==k){
                sum+=n->data;
            }
            if(n->left){
                q.push(n->left);
            }
            if(n->right){
                q.push(n->right);
            }
        }
        else if(!q.empty()){
                q.push(NULL);
                level++;
        }
    }
    return sum;
}


int countNode(node * root){
    if(root==NULL){
        return 0;
    }

    return countNode(root->left)+countNode(root->right)+1;
}

int sumAllNodes(node * root){
    if(root== NULL){
        return 0;
    }
    
    return sumAllNodes(root->left)+sumAllNodes(root->right)+root->data;
}

int heightOfTree(node * root){

    if(root== NULL){
        return 0;
    }

    int lHeight = heightOfTree(root->left);
    int rHeight = heightOfTree(root->right);
    return max(lHeight,rHeight)+1;
}


int calcuDiameter(node * root){

    //time complexcity = O(n^2)
    if(root==NULL){
        return 0;
    }

    int lHeight = heightOfTree(root->left);
    int rHeight = heightOfTree(root->right);
    int currDiameter = lHeight+rHeight+1;

    int lDiameter = calcuDiameter(root->left);
    int rDiameter = calcuDiameter(root->right);

    return max(currDiameter,max(lDiameter,rDiameter));

}

int calcuDiameter2(node * root,int * height){

    //time complexcity = O(n)
    if(root==NULL){
        *height = 0;
        return 0;
    }
    int  lh=0; int rh=0;

    int lDiameter = calcuDiameter2(root->left, &lh);
    int rDiameter = calcuDiameter2(root->right, &rh);

    int currDiameter = lh+rh+1;
    *height = max(lh,rh)+1;

    return max(currDiameter,max(lDiameter,rDiameter));

}



void sumReplace(node * root){
    if(root ==NULL){
        return;
    }
    
    sumReplace(root->left);
    sumReplace(root->right);

    if(root->left != NULL){
        root->data += root->left->data;
    }
    if(root->right != NULL){
        root->data += root->right->data;
    }
}


bool isBalanced(node * root){
    //time complexcity = O(n^2)
    if (root==NULL)
    {
        return true;
    }
    if(isBalanced(root->left)== false){
        return false;
    }
    if(isBalanced(root->right)== false){
        return false;
    }

    int lHeight = heightOfTree(root->left);
    int rHeight = heightOfTree(root->right);

    if(abs(lHeight-rHeight)<=1){
        return true;
    }else{
        return false; 
    }

}


bool isBalanced2(node * root,int * height){
    //time complexcity = O(n^2)
    int lh=0,rh =0;
    if (root==NULL)
    {
        return true;
    }
    if(isBalanced2(root->left,&lh)== false){
        return false;
    }
    if(isBalanced2(root->right,&rh)== false){
        return false;
    }

    *height = max(lh,rh)+1;
    
    if(abs(lh-rh)<=1){
        return true;
    }else{
        return false; 
    }

}


void reightView(node * root){
    if(root==NULL){
        return;
    }

    queue<node*>q;
    q.push(root);

    while (!q.empty())
    {
        int n = q.size();

        for(int i =0;i<n;i++){
            node * curr = q.front();
            q.pop();

            if(i==n-1){
                cout<<curr->data<<" ";
            }
            if(curr->left){
                q.push(curr->left);
            }
            if(curr->right){
                q.push(curr->right);
            }
        }
    }
    
}

void leftView(node * root){
    if(root==NULL){
        return;
    }

    queue<node*>q;
    q.push(root);

    while (!q.empty())
    {
        int n = q.size();

        for(int i =0;i<n;i++){
            node * curr = q.front();
            q.pop();

            if(i==0){
                cout<<curr->data<<" ";
            }
            if(curr->left){
                q.push(curr->left);
            }
            if(curr->right){
                q.push(curr->right);
            }
        }
    }
    
}


node * lca(node* root ,int n1, int n2){
    //lowest common ancestor

    if(root == NULL){
        return NULL;
    }
    if(root->data == n1 || root -> data == n2){
        return root;
    }

    node * left = lca(root->left,n1,n2);
    node * right = lca(root->right,n1,n2);

    if( left != NULL && right != NULL){
        return root;
    }
    if(left == NULL && right == NULL){
        return NULL;
    }
    if(left != NULL){
        return lca(root->left,n1,n2);
    }
    return lca(root->right,n1,n2);
}

int findDis(node * root ,int k,int dist){
    if(root==NULL){
        return -1;
    }
    if(root->data == k){
        return dist;
    }
    int left = findDis(root->left,k,dist+1);

    if(left != -1){
        return left;
    }

    return findDis(root->right,k,dist+1);
}


int distBtwnNode(node * root,int n1,int n2){

    node * LCA= lca(root,n1,n2);
    int d1 = findDis(LCA,n1,0);
    int d2 = findDis(LCA,n2,0);

    return d1+d2;
}



int main(){

    node* root = new node(1);

    root->left = new node(5);
    root->right = new node(3);

    root->left->left = new node(6);
    root->right->left = new node(2);

    root->left->right = new node(7);
    root->right->right = new node(4);

    preOrder(root);
    cout<<endl;

    inOrder(root);
    cout<<endl;
    postOrder(root);
    cout<<endl<<endl;
    levelOrder(root);
    cout<<endl<<endl;
    cout<<sumAtK(root,2)<<endl;
    cout<<countNode(root)<<endl;
    cout<<sumAllNodes(root)<<endl;
    cout<<heightOfTree(root)<<endl;
    cout<<calcuDiameter(root)<<endl;

    int height =0;
    cout<<calcuDiameter2(root, &height)<<endl;

    preOrder(root);
    cout<<endl;
    sumReplace(root);
    cout<<endl;
    preOrder(root);
    cout<<endl;

    if(isBalanced(root)){
        cout<<"tree is balanced "<<endl;
    }else{
        cout<<"tree is not balanced "<<endl;
    }

    int height2 =0;
    if(isBalanced2(root,&height2)){
        cout<<"tree is balanced "<<endl;
    }else{
        cout<<"tree is not balanced "<<endl;
    }
    cout<<endl;
    reightView(root);
    cout<<endl;
    leftView(root);
    cout<<endl;
    cout<<distBtwnNode(root,18,2)<<endl;

    return 0;
}

File: JAVA/tree/tree.java
import java.util.*;;

public class tree {
    
    static class node{
        int data;
        node left;
        node right;

        node(int val){
            this.data = val;
            this.left = null;
            this.right = null;
        }
    }

    public static void preorder(node head){
        if(head==null){
            return;
        }
        System.out.print(head.data+" ");
        preorder(head.left);
        preorder(head.right);
    }

    public static void inorder(node head){
        if(head==null){
            return;
        }
        inorder(head.left);
        System.out.print(head.data+" ");
        inorder(head.right);
    }

    public static void postorder(node head){
        if(head==null){
            return;
        }
        postorder(head.left);
        postorder(head.right);
        System.out.print(head.data+" ");
    }


    public static void leveOrder(node head){
        if(head == null){
            return ;
        }

        Queue<node> q = new LinkedList<>();
        q.add(head);
        q.add(null);

        while(!q.isEmpty()){
            node curr = q.peek();
            q.remove();
            if(curr!=null){
                System.out.print(curr.data+" ");
                if(curr.left != null){
                    q.add(curr.left);
                }
                if(curr.right != null){
                    q.add(curr.right);
                }
            }else if(!q.isEmpty()){
                q.add( null);
            }
        }
    }


    public static int sumAtKthLevel(node head,int k){
        if(head==null){
            return -1;
        }

        Queue<node> q = new LinkedList<>();

        q.add(head);
        q.add(null);
        int level = 0;
        int sum = 0;

        while(!q.isEmpty()){

            node curr = q.peek();
            q.remove();
            if(curr!=null){
                if(level==k){
                    sum+=curr.data;
                }
                if(curr.left != null){
                    q.add(curr.left);
                }
                if(curr.right != null){
                    q.add(curr.right);
                }
            }else if(!q.isEmpty()){
                level++;
                q.add( null);
            }
        }
        return sum;
    }

    public static int totalNodes(node head){
        if(head==null){
            return 0;
        }
        return totalNodes(head.left)+totalNodes(head.right)+1;
    }

    public static int sumNode(node head){
        if(head == null){
            return 0;
        }

        return sumNode(head.left)+sumNode(head.right)+head.data;
    }

    public static int heightOfTree(node head){
        if(head == null){
            return 0;
        }

        int lHeight = heightOfTree(head.left);
        int rHeight = heightOfTree(head.right);

        return Math.max(lHeight,rHeight)+1;
    }

    public static int diameterOfTree(node head){
        if(head == null){
            return 0;
        }

        int lHeight = heightOfTree(head.left);
        int rHeight = heightOfTree(head.right);
        int currDiameter = lHeight+rHeight+1;

        int lDiameter = diameterOfTree(head.left);
        int rDiameter = diameterOfTree(head.right);
        
        return Math.max(currDiameter, Math.max(lDiameter,rDiameter));
    }

    public static void sumReplace(node head){
        if(head == null){
            return;
        }

        sumReplace(head.left);
        sumReplace(head.right);
        if(head.left != null){
            head.data+=head.left.data;
        }
        if(head.right != null){
            head.data+=head.right.data;
        }
    }

    public static boolean isBalancedTree(node head){

        if(head == null){
            return true;
        }

        if(isBalancedTree(head.left)==false){
            return false;
        }

        if(isBalancedTree(head.right)==false){
            return false;
        }

        int lHeight = heightOfTree(head.left);
        int rHeight = heightOfTree(head.right);

        if(Math.abs(lHeight-rHeight)<=1){
            return true;
        }else{
            return false;
        }
    }

    public static void rightView(node root){
        if(root == null){
            return;
        }
        
        Queue<node> q=new LinkedList<>();
        q.add(root);

        while(!q.isEmpty()){
            int n = q.size();

            for(int i=0;i<n;i++){

                node curr = q.peek();
                q.remove();
                if(i==n-1){
                    System.out.print(curr.data+" ");
                }
                if(curr.left != null){
                    q.add(curr.left);
                }
                if(curr.right != null){
                    q.add(curr.right);
                }
            }
        }
            
    }

    public static void leftView(node root){
        if(root == null){
            return;
        }
        
        Queue<node> q=new LinkedList<>();
        q.add(root);

        while(!q.isEmpty()){
            int n = q.size();

            for(int i=0;i<n;i++){

                node curr = q.peek();
                q.remove();
                if(i==0){
                    System.out.print(curr.data+" ");
                }
                if(curr.left != null){
                    q.add(curr.left);
                }
                if(curr.right != null){
                    q.add(curr.right);
                }
            }
        }
            
    }


    public static void main(String args[]) {

        node head = new node(4);
        head.left=new node(6);
        head.right =new node(8);
        head.left.left = new node(5);
        head.left.right = new node(2);
        head.right.left = new node(9);
        head.right.right = new node(1);
        head.left.left.left = new node(3);

        // preorder(head);
        // inorder(head);
        // postorder(head);
        // leveOrder(head);
        // System.out.println(sumAtKthLevel(head, 2));
        // System.out.println(sumNode(head));
        // System.out.println(totalNodes(head));

        // System.out.println(heightOfTree(head));
        // System.out.println(diameterOfTree(head));

        // preorder(head);
        // System.out.println();
        // sumReplace(head);
        // preorder(head);

        // if(isBalancedTree(head)){
        //     System.out.println("Tree is Balanced");
        // }else{
        //     System.out.println("tree is not balanced");
        // }
        
        // rightView(head);
        // System.out.println();
        // leftView(head);
        
    }
}
